File: 0040-combination-sum-ii/solution.cpp
class Solution {
public:
    void backtrack(vector<vector<int>>& ans, vector<int>& nums, int target, int start, vector<int> res) {
        if (target == 0) {
            ans.push_back(res);
            return;
        }
        
        for (int i = start; i < nums.size(); i++) {
            if (i > start && nums[i] == nums[i - 1]) continue;
            if (nums[i] > target) continue;
            res.push_back(nums[i]);
            backtrack(ans, nums, target - nums[i], i+1, res);
            res.pop_back();
        }
    }
    
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        sort(candidates.begin(),candidates.end());
        backtrack(ans, candidates, target, 0, {});  
        return ans;
    }
};

File: 0040-combination-sum-ii/solution.java
class Solution {
    void logic(int ind,int[] candidates,int target,ArrayList<Integer> temp, List<List<Integer>> ans){
        if(target==0){
            ans.add(new ArrayList<>(temp));
            return;
        }

        for(int i=ind;i<candidates.length;i++){
            if(i>ind && candidates[i-1]==candidates[i]) continue;
            if(candidates[i]>target) continue;
            temp.add(candidates[i]);
            logic(i+1,candidates,target-candidates[i],temp,ans);
            temp.remove(temp.size()-1);
        }

    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> ans=new ArrayList<>();
        logic(0,candidates,target,new ArrayList<>(),ans);
        return ans;
    }
}
